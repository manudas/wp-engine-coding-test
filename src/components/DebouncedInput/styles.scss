.debounced-input {
    &_input {
        display: block;
        width: 100px;
    }
}