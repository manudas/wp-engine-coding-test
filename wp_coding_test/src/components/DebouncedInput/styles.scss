@import '../globals';

.debounced-input {
    &__input {
        display: block;
        width: 100px;

        border-radius: 3px;
        border: 1px solid $semiTransparentBlack;
        background-color: $semiTransparentWhite2;

        color: $darkerGrey;
    }
}