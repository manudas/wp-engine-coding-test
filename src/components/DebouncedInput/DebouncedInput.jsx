import React, {
    useState,
    useCallback,
} from 'react';

import debounce from 'lodash/debounce';

import {
    debounceTimeout,
    defaultReddit,
} from '../constants';

import './styles.scss';

const DebouncedInput = ({
    onChangeHandler,
}) => {
    const [ subReddit, setSubReddit ] = useState(defaultReddit);

    const debouncedHandler = useCallback(debounce(
        (passedReddit) => {
            // console.log('debounced!');
            onChangeHandler(passedReddit);
        },
        debounceTimeout,
        {
            'leading': false,
            'trailing': true,
        }), []);

    const onInputChangeHandler = ($event) => {
        setSubReddit($event.target.value);
        debouncedHandler($event.target.value); 
    };

    return (
        <label
            className="debounced-input_label"
            htmlFor="subReddit">
                Please enter a subreddit
                <input
                    className="debounced-input_input"
                    onChange={onInputChangeHandler}
                    id="subreddit"
                    value={subReddit}
                />
        </label>
    );
};

export default DebouncedInput;
