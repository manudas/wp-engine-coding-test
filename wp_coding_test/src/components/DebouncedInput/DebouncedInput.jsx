import React, {
    useState,
    useMemo,
} from 'react';

import debounce from 'lodash/debounce';

import {
    debounceTimeout,
    defaultReddit,
} from '../constants';

import './styles.scss';

const DebouncedInput = ({
    onChangeHandler,
}) => {
    const [ subReddit, setSubReddit ] = useState(defaultReddit);

    const debouncedHandler = useMemo( () => debounce(
        (passedReddit) => {
            // console.log('debounced!');
            onChangeHandler(passedReddit);
        },
        debounceTimeout,
        {
            'leading': false,
            'trailing': true,
        }), [onChangeHandler]);

    const onInputChangeHandler = ($event) => {
        setSubReddit($event.target.value);
        debouncedHandler($event.target.value);
    };

    return (
        <label
            className="debounced-input_label"
            htmlFor="subreddit">
                Please enter a subreddit
                <input
                    data-testid="debouncedInput"
                    className="debounced-input_input"
                    onChange={onInputChangeHandler}
                    id="subreddit"
                    value={subReddit}
                />
        </label>
    );
};

export default DebouncedInput;
