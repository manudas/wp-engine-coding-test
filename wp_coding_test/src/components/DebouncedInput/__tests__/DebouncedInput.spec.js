import React from 'react';
import {
    fireEvent,
    render,
} from '@testing-library/react';
import DebouncedInput from '..';

describe('DebouncedInput test suite', () => {

    let handler;
    let renderer;

    // long-runing test
    jest.setTimeout(10000); // 10 secs

    beforeEach(() => {
        handler = jest.fn();
        renderer = render(
            <DebouncedInput
                onChangeHandler={handler}
            />
        );
    });

    it('should render <DebouncedInput />', () => {
        const { container } = renderer;
        expect(container).toMatchSnapshot();
    });

    it('should change input value and should be debounced', async () => {
        const sleep = (ms) => {
            return new Promise(resolveFn => setTimeout(resolveFn, ms));
        }

        const { getByTestId } = renderer;
        const input = getByTestId('debouncedInput');

        // default value is reactjs, which is 7 chars long
        const defaultValue = 'reactjs';
        for (let index = 0; index < defaultValue.length; index++) {
            // lets simulate we are deleting the content of the input
            fireEvent.change(
                input,
                {
                    target:
                        {
                            value: defaultValue.substring(0, (defaultValue.length - 1) - index)
                        }
                }
            );
        }

        const longSubredditName = 'UNBGBBIIVCHIDCTIICBG';
        const timeBeetweenKeystrokes = 200; // ms
        for (let index = 0; index < longSubredditName.length; index++) {

            /* lets simulate we are typing a long
             * subreddit name with a delay of 220
             * milliseconds between each key stroke
             */
            fireEvent.change(
                input,
                {
                    target:
                        {
                            value: longSubredditName.substring(0, index + 1)
                        }
                }
            );
            // debounceTimeout is 500
            await sleep(timeBeetweenKeystrokes);
        }

        // the final result should be the long subreddit name
        expect(input.value).toBe(longSubredditName);

        // debounceTimeout is 500
        // let's let the function to be run
        await sleep(600);
        // as it's debounced, we should only have called it once
        expect(handler).toHaveBeenCalledTimes(1);
    });
});
