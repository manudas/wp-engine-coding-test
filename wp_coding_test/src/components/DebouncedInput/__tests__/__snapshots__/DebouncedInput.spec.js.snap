// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DebouncedInput test suite should render <DebouncedInput /> 1`] = `
<div>
  <label
    class="debounced-input_label"
    for="subreddit"
  >
    Please enter a subreddit
    <input
      class="debounced-input_input"
      data-testid="debouncedInput"
      id="subreddit"
      value="reactjs"
    />
  </label>
</div>
`;
